test/metabase/query_processor/middleware/metrics_test.clj

```diff
@@ -0,0 +1,119 @@
+(ns metabase.query-processor.middleware.metrics-test
+  (:require
+   [clojure.test :refer [deftest is]]
+   [metabase.lib.convert :as lib.convert]
+   [metabase.lib.core :as lib]
+   [metabase.lib.equality :as lib.equality]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.lib.metadata.jvm :as lib.metadata.jvm]
+   [metabase.lib.options :as lib.options]
+   [metabase.lib.test-metadata :as meta]
+   [metabase.lib.test-util :as lib.tu]
+   [metabase.query-processor :as qp]
+   [metabase.query-processor.middleware.metrics :as metrics]
+   [metabase.test :as mt]))
+
+(def counter (atom 2000))
+
+(defn- mock-metric
+  ([]
+   (mock-metric meta/metadata-provider identity))
+  ([f]
+   (mock-metric meta/metadata-provider f))
+  ([metadata-provider f]
+   (let [query (-> (lib/query meta/metadata-provider (meta/table-metadata :products))
+                   (lib/aggregate (lib/avg (meta/field-metadata :products :rating)))
+                   f)
+         metric {:lib/type :metadata/card
+                 :id (swap! counter inc)
+                 :database-id (meta/id)
+                 :name "Mock metric - Avertage Product Rating"
+                 :type :model
+                 :dataset-query query}]
+     [metric (lib/composed-metadata-provider
+               metadata-provider
+               (lib.tu/mock-metadata-provider
+                 {:cards [metric]}))])))
+
+(defn- add-metric-source [query source-metric]
+  (-> query
+      (update-in [:stages 0] (fn [stage]
+                               (-> stage
+                                   (assoc :sources [{:lib/type :source/metric :id  (:id source-metric)}])
+                                   (dissoc :source-card :source-table))))
+      (lib/aggregate [:metric {} (:id source-metric)])))
+
+(deftest expand-aggregation-metric-ref-test
+  (let [[source-metric mp] (mock-metric)
+        query (-> (lib/query mp source-metric)
+                  (add-metric-source source-metric)
+                  (lib/aggregate (lib/+ (lib.options/ensure-uuid [:metric {} (:id source-metric)]) 1)))]
+    (is (=?
+          {:stages [{:source-table (meta/id :products)
+                     :aggregation [[:avg {} [:field {} (meta/id :products :rating)]]
+                                   [:+ {} [:avg {} [:field {} (meta/id :products :rating)]] 1]]}]}
+          (metrics/expand query)))))
+
+(deftest expand-basic-test
+  (let [[source-metric mp] (mock-metric)
+        query (-> (lib/query mp source-metric)
+                  (add-metric-source source-metric))]
```

Comment:
This feels like a repeated block that's worth a helper function around `(mock-metric)`. Like:
```clojure
(defn- metric-query [& args]
  (let [[source-metric mp] (apply mock-metric args)]
    (-> (lib/query mp source-metric)
        (add-metric-source source-metric))))
```

Reply:
Oh, with maybe an optional function to go between the query and adding the metric aggregation, since I see that a bunch of things add filters etc. before it and the order might matter.

Reply:
`add-metric-source` should get wiped out once the `lib` functions exist to build these queries.